### PR TITLE
Lock fixings & unittests

### PR DESCRIPTION
**Pull request by nupplaphil**
- Added Redis Lock Unittests
- Added Memcached Lock Unittests

- Fixed a bug in dba (index error) - @annando was that right?
- Fixed bugs in RedisLock because of the unittests ;-)

- Changed Redis "serialize" to "json_encode" because of issues with integer/boolean conversions => no problems with json

### CODE REVIEW

**Comment by MrPetovan** [comment]
Hooray for unit tests!

**Comment by nupplaphil** [comment]
Next station: Cache Unit-Tests (I already prepared some, but because of the bugs in redis, I pushed this one first)

**Comment by nupplaphil** [comment]
**Please don't merge it yet!** I found some issues because of the cache unittests 
=> I'll merge them all and will tell you when I'm done

**Comment by MrPetovan** [comment]
You can close this PR and reopen it when you’re ready.

File: tests/src/Core/Cache/CacheTest.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		Config::set('system', 'throttle_limit_day', 100);
+		Config::set('system', 'throttle_limit_week', 100);
+		Config::set('system', 'throttle_limit_month', 100);
+		Config::set('system', 'theme', 'system_theme');
+	}
+
+	function testSimple() {
+		$this->assertNull($this->instance->get('value1'));
+
+		$value='foobar';
```
Comment (MrPetovan) [inline tests/src/Core/Cache/CacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/CacheTest.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		$this->instance->set('value1', $value);
+		$received=$this->instance->get('value1');
+		$this->assertEquals($value, $received, 'Value received from cache not equal to the original');
+		$value='ipsum lorum';
```
Comment (MrPetovan) [inline tests/src/Core/Cache/CacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/CacheTest.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		$this->instance->set('value1', $value);
+		$received=$this->instance->get('value1');
```
Comment (MrPetovan) [inline tests/src/Core/Cache/CacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/CacheTest.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		$this->assertEquals($value, $received, 'Value not overwritten by second set');
+
+		$value2='foobar';
```
Comment (MrPetovan) [inline tests/src/Core/Cache/CacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/CacheTest.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+			'1_value2' => $this->instance->get('1_value2'),
+			'2_value1' => $this->instance->get('2_value1'),
+			'3_value1' => $this->instance->get('3_value1'),
+		]);
+	}
+
+	function testTTL() {
+		$this->assertNull($this->instance->get('value1'));
+
+		$value='foobar';
```
Comment (MrPetovan) [inline tests/src/Core/Cache/CacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/CacheTest.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		$this->instance->set('value1', $value, 1);
+		$received=$this->instance->get('value1');
```
Comment (MrPetovan) [inline tests/src/Core/Cache/CacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/MemoryCacheTest.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		parent::setUp();
+		if (!($this->instance instanceof IMemoryCacheDriver)) {
+			throw new \Exception('MemoryCacheTest unsupported');
+		}
+	}
+
+	function testCompareSet() {
+		$this->assertNull($this->instance->get('value1'));
+
+		$value='foobar';
```
Comment (MrPetovan) [inline tests/src/Core/Cache/MemoryCacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/MemoryCacheTest.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		$this->instance->add('value1', $value);
+		$received=$this->instance->get('value1');
```
Comment (MrPetovan) [inline tests/src/Core/Cache/MemoryCacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/MemoryCacheTest.php [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		$this->assertEquals($value, $received, 'Value received from cache not equal to the original');
+		$newValue='ipsum lorum';
```
Comment (MrPetovan) [inline tests/src/Core/Cache/MemoryCacheTest.php]: Code standards: Please wrap operators with spaces.

File: tests/src/Core/Cache/MemoryCacheTest.php [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		$this->instance->compareSet('value1', $value, $newValue);
+		$received=$this->instance->get('value1');
```
Comment (MrPetovan) [inline tests/src/Core/Cache/MemoryCacheTest.php]: Code standards: Please wrap operators with spaces.

**Comment by nupplaphil** [comment]
hm .. OK Travis doesn't support memcache for PHP7.X , just memcached..
I'm disabling the support in the .travis.yml but you can use these unittests at local environments (like I do)

**Comment by nupplaphil** [comment]
Something strange is happening sometimes:

I'm getting randomly index errors like
- `Notice: Undefined index: scheme in /home/travis/build/nupplaphil/friendica/src/App.php on line 417`
- `Notice: Undefined index: host in /home/travis/build/nupplaphil/friendica/src/App.php on line 419`
- `Notice: Undefined variable: ret in /home/travis/build/nupplaphil/friendica/include/dba.php on line 971`

I already fixed one in the `dba.php` with this PR. @annando is it possible that there are bugs because lack of "key-checks" in these kind of arrays? It seems like for me

**Comment by nupplaphil** [comment]
https://travis-ci.org/nupplaphil/friendica/jobs/401278562#L563
https://travis-ci.org/nupplaphil/friendica/jobs/401278562#L563

**Comment by MrPetovan** [comment]
See here for the scheme and host issue: https://github.com/friendica/friendica/pull/5295#issuecomment-403190489

**Comment by nupplaphil** [comment]
btw: I'm done with my changes :-)

**Comment by annando** [comment]
Will this fix the currently failing tests?

**Comment by nupplaphil** [comment]
Don't think that it will fix everything. I think we have to search for empty key access failures in  `dba.php` and `App.php`. I fixed it at the place where I found one.

But I was afraid of 'fixing' something which was for purpose so I didn't search every occurence of it. Because you're currently working hard on this part of code :-)

**Pull request merged**